projects/saturn/src/main/java/io/fairspace/saturn/services/views/MaterializedViewService.java

```diff
@@ -37,47 +38,109 @@ public void createOrUpdateAllMaterializedViews() {
+            // all checks and changes to be done in one transaction
+            connection.setAutoCommit(false);
+            try {
+                List<String> columns = collectViewColumns(view);
+                if (doesMaterializedViewExist(mvName, connection)) {
+                    // 'refresh' requires unique index based on all columns, so we have to check if it exists
+                    createMaterializedViewUniqueIndexIfNotExist(viewName + UNIQUE_INDEX_POSTFIX, mvName, columns, connection);
+                    refreshMaterializedView(mvName, connection);
+                } else {
+                    createViewMaterializedView(mvName, view, setColumns, connection);
+                    createMaterializedViewIndex(viewName + INDEX_POSTFIX, mvName, viewName + "id", connection);
+                    createMaterializedViewUniqueIndexIfNotExist(viewName + UNIQUE_INDEX_POSTFIX, mvName, columns, connection);
+                }
+                connection.commit();
+                log.info("View materialized view {} refresh has finished successfully", mvName);
+            } catch (SQLException e) {
+                connection.rollback();
+                throw e;
+            } finally {
+                connection.setAutoCommit(true);
+            }
         }
     }
 
-    public void createOrUpdateJoinMaterializedView(ViewsConfig.View view, Connection connection) throws SQLException {
+    private void createOrUpdateJoinMaterializedView(ViewsConfig.View view, Connection connection) throws
+            SQLException {
         for (ViewsConfig.View.JoinView joinView : view.join) {
             String viewName = view.name.toLowerCase();
             String joinViewName = joinView.view.toLowerCase();
             var mvName = "mv_%s_join_%s".formatted(viewName, joinViewName);
-            log.info("{} refresh has started", mvName);
-            dropMaterializedViewIfExists(mvName, connection);
-            createJoinMaterializedViews(view, joinView, connection);
-            createMaterializedViewIndex(mvName + "_" + viewName + INDEX_POSTFIX, mvName, viewName + "_id", connection);
-            createMaterializedViewIndex(
-                    mvName + "_" + joinViewName + INDEX_POSTFIX, mvName, joinViewName + "_id", connection);
-            log.info("{} refresh has finished successfully", mvName);
+
+            log.info("Join materialized view {} refresh has started", mvName);
+            // all checks and changes to be done in one transaction
+            connection.setAutoCommit(false);
+            try {
+                List<String> columns = collectJoinColumns(view, joinView);
+                if (doesMaterializedViewExist(mvName, connection)) {
+                    // 'refresh' requires unique index based on all columns, so we have to check if it exists
+                    createMaterializedViewUniqueIndexIfNotExist(mvName + UNIQUE_INDEX_POSTFIX, mvName, columns, connection);
```

Comment:
so the unique index doesn't need to be commited before refresing materialized view? Is it enough that you do 'ps.execute()'?

Reply:
yes, it is visible in the scope of one transaction